src/components/tree/Tree.tsx

```diff
@@ -0,0 +1,152 @@
+
+    // Check if the current catalog has sub-catalogs or collections
+    const subCatalogs: TreeCatalog[] = await Promise.all(
+      catalogData.links
+        ?.filter((link) => link.rel === 'child')
+        .map((link: Link) => fetchCatalog(link.href)) || [],
+    );
+
+    const collectionUrl = catalogData.links
+      .filter((link) => link.rel === 'data')
+      .filter((link) => link.href.includes('/collections'))
+      .map((link) => link.href)[0];
+
+    let collections: Collection[] = [];
+    if (collectionUrl) {
+      const collectionsResponse = await fetchData(collectionUrl);
+      collections = collectionsResponse.collections;
+    }
+
+    // Return the catalog with its children (sub-catalogs and collections)
+    return {
+      ...catalogData,
+      catalogs: subCatalogs,
+      collections,
+    };
+  }, []);
+
+  // Fetch the ROOTS array of URLs when the component mounts
+  useEffect(() => {
+    const fetchRoots = async () => {
+      const rootsResponse = await fetchData(catalogUrl);
+
+      const rootCatalogUrls = rootsResponse.catalogs.reduce((acc, obj) => {
+        const selfLink = obj.links.find((link: Link) => link.rel === 'self');
+
+        if (selfLink) {
+          acc.push(selfLink.href);
+        }
+
+        return acc;
+      }, []);
+
+      const topLevelCatalogs = await Promise.all(
+        rootCatalogUrls.map((url: string) => fetchCatalog(url)),
+      );
+      setTreeData(topLevelCatalogs);
+    };
+
+    fetchRoots();
+  }, [catalogUrl, fetchCatalog]);
+
+  // Recursive function to filter tree nodes based on the filter text
+  const filter = useCallback(
+    (node: TreeCatalog | Collection): TreeCatalog | Collection | null =>
+      filterTree(node, activeFilters),
+    [activeFilters],
+  );
+
+  // Debounce the filtering of the tree
+  useDebounce(
+    () => {
+      if (treeData.length > 0) {
+        const filteredData = treeData
+          .map((catalog) => filter(catalog))
+          .filter(Boolean) as TreeCatalog[];
+        setFilteredTreeData(filteredData);
+      }
+    },
+    300,
+    [treeData, activeFilters, filter],
+  );
+
+  // Toggle the expansion state of a catalog node
+  const toggleExpand = (nodeId: string) => {
+    setExpandedNodes((prevState) => ({
+      ...prevState,
+      [nodeId]: !prevState[nodeId], // Toggle the expanded state of this node
+    }));
+  };
+
+  const handleLeafClick = (node: Collection) => {
+    setSelectedCollection(node);
+    setActivePage('items');
+  };
+
+  return (
+    <div>
+      {filteredTreeData.length > 0 ? (
+        <ul className="branch">
+          {filteredTreeData.map((catalog) => (
+            <TreeNode
+              key={catalog.id}
+              expandedNodes={expandedNodes}
+              handleLeafClick={handleLeafClick}
+              node={catalog}
+              toggleExpand={toggleExpand}
+            />
+          ))}
+        </ul>
+      ) : (
+        <div>Loading...</div>
```

Comment:
Could be misleading. Based on the work you showed me a few weeks ago, we can put a root catalog URL as a param, so if there is no sub catalogs inside of that path, then it would be seen as Loading... indefinitely

Reply:
Good point, hadn't thought of that, I think maybe it be better to set the filtered tree data to `null` so that it hasn't yet been fetched, as you say a response of an empty array is a valid response, so if I only show the loading while it is null, it should satisfy both states.

Of course if you have a different suggestion, feel free, but I'll definitely look into this.

Reply:
Or just a separate loading state

Reply:
I don't think so, it is always better to infer state from other data if possible and the tree data is only ever not null if we successfully fetched data .